File: apps/api/app/schemas/retrieval.py
# Bu sema dosyasi, retrieval icin API veri sozlesmelerini tanimlar.

from __future__ import annotations

from typing import Any, Literal

from pydantic import BaseModel, Field


class RetrievalHints(BaseModel):
    keywords: list[str] = Field(default_factory=list)
    section_tags: list[str] = Field(default_factory=list)
    period: str | None = None
    small_to_big: bool = False
    context_window: int = Field(default=1, ge=0, le=3)


class RetrievalQueryRequest(BaseModel):
    tenant_id: str = Field(min_length=1)
    project_id: str = Field(min_length=1)
    query_text: str = Field(min_length=2)
    top_k: int = Field(default=10, ge=1, le=50)
    retrieval_mode: Literal["hybrid", "sparse", "dense"] = "hybrid"
    min_score: float = Field(default=0.0, ge=0.0)
    min_coverage: float = Field(default=0.0, ge=0.0, le=1.0)
    retrieval_hints: RetrievalHints | None = None


class EvidenceResult(BaseModel):
    evidence_id: str
    source_document_id: str
    chunk_id: str
    page: int | None
    text: str
    score_dense: float | None
    score_sparse: float | None
    score_final: float
    metadata: dict[str, Any]


class RetrievalDiagnostics(BaseModel):
    backend: str
    retrieval_mode: str
    top_k: int
    result_count: int
    filter_hit_count: int
    coverage: float = Field(ge=0.0, le=1.0)
    best_score: float = Field(ge=0.0)
    quality_gate_passed: bool
    latency_ms: int
    index_name: str
    applied_filters: dict[str, str]


class RetrievalQueryResponse(BaseModel):
    retrieval_run_id: str
    evidence: list[EvidenceResult]
    diagnostics: RetrievalDiagnostics

File: apps/web/src/lib/api/retrieval.ts
import { useMutation } from "@tanstack/react-query";
import { z } from "zod";

import type { WorkspaceContext } from "./client";
import { createOpenApiClient, parseOpenApiResult, workspaceHeaders } from "./core";
import { jsonObjectSchema } from "./schema-helpers";

export const retrievalModeSchema = z.enum(["hybrid", "sparse", "dense"]);

export const retrievalLabFormSchema = z.object({
  queryText: z.string().trim().min(2, "Query must be at least 2 characters."),
  topK: z.string().trim().default("10"),
  retrievalMode: retrievalModeSchema,
  minScore: z.string().trim().default("0"),
  minCoverage: z.string().trim().default("0"),
  period: z.string().trim().default(""),
  keywords: z.string().trim().default(""),
  sectionTags: z.string().trim().default(""),
});

export const retrievalRequestSchema = z.object({
  tenant_id: z.string().trim().min(1),
  project_id: z.string().trim().min(1),
  query_text: z.string().trim().min(2),
  top_k: z.number().int().positive(),
  retrieval_mode: retrievalModeSchema,
  min_score: z.number(),
  min_coverage: z.number(),
  retrieval_hints: z.object({
    period: z.string().trim().nullable(),
    keywords: z.array(z.string()),
    section_tags: z.array(z.string()),
    small_to_big: z.boolean(),
    context_window: z.number().int().nonnegative(),
  }),
});

const evidenceItemSchema = z.object({
  evidence_id: z.string(),
  source_document_id: z.string(),
  chunk_id: z.string(),
  page: z.number().nullable(),
  text: z.string(),
  score_dense: z.number().nullable(),
  score_sparse: z.number().nullable(),
  score_final: z.number(),
  metadata: jsonObjectSchema,
});

export const retrievalResponseSchema = z.object({
  retrieval_run_id: z.string(),
  evidence: z.array(evidenceItemSchema),
  diagnostics: z.object({
    backend: z.string(),
    retrieval_mode: z.string(),
    top_k: z.number(),
    result_count: z.number(),
    filter_hit_count: z.number(),
    coverage: z.number(),
    best_score: z.number(),
    quality_gate_passed: z.boolean(),
    latency_ms: z.number(),
    index_name: z.string(),
    applied_filters: z.record(z.string(), z.string()),
  }),
});

export type RetrievalResponse = z.infer<typeof retrievalResponseSchema>;
export type RetrievalLabFormInput = z.infer<typeof retrievalLabFormSchema>;

export function buildRetrievalRequest(
  workspace: WorkspaceContext,
  input: RetrievalLabFormInput,
) {
  const parsed = retrievalLabFormSchema.parse(input);

  return retrievalRequestSchema.parse({
    tenant_id: workspace.tenantId,
    project_id: workspace.projectId,
    query_text: parsed.queryText,
    top_k: Number(parsed.topK) || 10,
    retrieval_mode: parsed.retrievalMode,
    min_score: Number(parsed.minScore) || 0,
    min_coverage: Number(parsed.minCoverage) || 0,
    retrieval_hints: {
      period: parsed.period.length > 0 ? parsed.period : null,
      keywords: parsed.keywords
        .split(",")
        .map((item) => item.trim())
        .filter(Boolean),
      section_tags: parsed.sectionTags
        .split(",")
        .map((item) => item.trim())
        .filter(Boolean),
      small_to_big: true,
      context_window: 1,
    },
  });
}

export async function queryRetrieval(
  workspace: WorkspaceContext,
  input: RetrievalLabFormInput,
): Promise<RetrievalResponse> {
  const client = createOpenApiClient();
  const body = buildRetrievalRequest(workspace, input);

  return parseOpenApiResult(
    await client.POST("/retrieval/query", {
      headers: workspaceHeaders({ workspace }),
      body,
    }),
    retrievalResponseSchema,
  );
}

export function useRetrievalQueryMutation(workspace: WorkspaceContext | null) {
  return useMutation({
    mutationFn: (input: RetrievalLabFormInput) => {
      if (!workspace) {
        throw new Error("Workspace not selected. Create/select workspace from New Report first.");
      }

      return queryRetrieval(workspace, input);
    },
  });
}
